### PR TITLE
MAINT: fix SHGO extra arguments

### LINKED ISSUE #14589: `shgo` error since scipy 1.8.0.dev0+1529.803e52d
`shgo` raises an error with the following example since scipy 1.8.0.dev0+1529.803e52d.

This error was noticed in the hyperspy test suite (https://github.com/hyperspy/hyperspy-extensions-list/actions/runs/1111267440) and previous scipy development version (1.8.0.dev0+1503.8647ee9) didn't raise an error. Scipy development version are installed from https://pypi.anaconda.org/scipy-wheels-nightly/simple.

#### Reproducing code example:

```python
import sys

import numpy as np
import scipy
from scipy.optimize import shgo

print(scipy.__version__, np.__version__, sys.version_info)


def func(x_values, param):
    a, b = param
    return a*x_values + b


def errfunc(param, y):
    return func(x_values, param) - y


def errfunc_sq(param, y, weights=None):
    if weights is None:
        weights = 1.0
    return ((weights * errfunc(param, y)) ** 2).sum()

x_values = np.arange(100)
y_values = func(x_values, (2, 10)) + np.random.random(x_values.size)

results = shgo(errfunc_sq,
               bounds=((-100, 100), (-100, 100)),
               args=(y_values, None)
               )

print(results)
```

#### Error message:

```python
1.8.0.dev0+1529.803e52d 1.21.1 sys.version_info(major=3, minor=8, micro=10, releaselevel='final', serial=0)
Traceback (most recent call last):
  File "/opt/miniconda3/envs/test_env/lib/python3.8/site-packages/scipy/optimize/_shgo_lib/triangulation.py", line 630, in __getitem__
    return self.cache[x]
KeyError: (0, 0)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "untitled0.py", line 34, in <module>
    results = shgo(errfunc_sq,
  File "/opt/miniconda3/envs/test_env/lib/python3.8/site-packages/scipy/optimize/_shgo.py", line 420, in shgo
    shc.construct_complex()
  File "/opt/miniconda3/envs/test_env/lib/python3.8/site-packages/scipy/optimize/_shgo.py", line 733, in construct_complex
    self.iterate()
  File "/opt/miniconda3/envs/test_env/lib/python3.8/site-packages/scipy/optimize/_shgo.py", line 876, in iterate
    self.iterate_complex()
  File "/opt/miniconda3/envs/test_env/lib/python3.8/site-packages/scipy/optimize/_shgo.py", line 895, in iterate_hypercube
    self.HC = Complex(self.dim, self.func, self.args,
  File "/opt/miniconda3/envs/test_env/lib/python3.8/site-packages/scipy/optimize/_shgo_lib/triangulation.py", line 25, in __init__
    self.n_cube(dim, symmetry=symmetry)
  File "/opt/miniconda3/envs/test_env/lib/python3.8/site-packages/scipy/optimize/_shgo_lib/triangulation.py", line 76, in n_cube
    self.C0.add_vertex(self.V[origintuple])
  File "/opt/miniconda3/envs/test_env/lib/python3.8/site-packages/scipy/optimize/_shgo_lib/triangulation.py", line 634, in __getitem__
    xval = Vertex(x, bounds=self.bounds,
  File "/opt/miniconda3/envs/test_env/lib/python3.8/site-packages/scipy/optimize/_shgo_lib/triangulation.py", line 557, in __init__
    self.f = func(x_a, *func_args)
  File "/opt/miniconda3/envs/test_env/lib/python3.8/site-packages/scipy/optimize/optimize.py", line 464, in function_wrapper
    fx = function(np.copy(x), *(wrapper_args + args))
TypeError: errfunc_sq() takes from 2 to 3 positional arguments but 5 were given
```

#### Scipy/Numpy/Python version information:

1.8.0.dev0+1529.803e52d 1.21.1 sys.version_info(major=3, minor=8, micro=10, releaselevel='final', serial=0)

### PR DESCRIPTION
**Pull request by o-alexandre-felipe**
#### Reference issue
Closes #14589

#### What does this implement/fix?
Passing unwrapped function to complex as suggested by jjramsey https://github.com/scipy/scipy/issues/14589#issuecomment-1096793518

#### Additional information
This case was not covered by the existing tests. Added a minimal failing test, and repeated, no failing tests after regression.

### CODE REVIEW

**Comment by WarrenWeckesser** [comment]
I'd rather give @jjramsey a chance to finish https://github.com/scipy/scipy/pull/15986, if @jjramsey is interested in doing so.  #16313 is still a draft; the fix in https://github.com/scipy/scipy/pull/15986 is simple, so it shouldn't cause any difficult conflicts when #16313 is resumed.

**Comment by tupui** [comment]
> I'd rather give @jjramsey a chance to finish #15986, if @jjramsey is interested in doing so. #16313 is still a draft; the fix in #15986 is simple, so it shouldn't cause any difficult conflicts when #16313 is resumed.

As you wish. I will let you handle this PR then.

I personally don't want to dissipate my efforts on the same thing. Most issues in `shgo` can be patched, but then we give efforts into something that we don't want to maintain moving forward. We have 6 months before the new release which is plenty if we focus and set this as a priority.

**Comment by o-alexandre-felipe** [comment]
Hi @tupui, yes I can help with some of those issues, but I don't want to spend too much time doing stuff not related to the goal of the contribution.

We should follow a test driven design approach. 
1. You, or someone else acquainted with the issue write unit tests with good coverage and a description in the docstring.
2. I, or some other developer interested on contributing, fork your branch and work on the fixing the issues revealed by the tests.
3. You review the pull request and merge.

Doing this we can speedup the review process, and encourage contributors by giving a clear problem statement and an objective goal.

I personally don't contribute very often different open source projects because I consider the overhead of setting up a development environment, debugging, and the possibility of submitting something that may be not accepted because done in the way the project administrators expected.

**Comment by tupui** [comment]
> We should follow a test driven design approach.
> 
> 1. You, or someone else acquainted with the issue write unit tests with good coverage and a description in the docstring.
> 2. I, or some other developer interested on contributing, fork your branch and work on the fixing the issues revealed by the tests.
> 3. You review the pull request and merge.
> 
> Doing this we can speedup the review process, and encourage contributors by giving a clear problem statement and an objective goal.

I am afraid, for such project, it does not work in practice. Most people are doing one shot contributions and things like strict xfail of a test are hard to spot/keep track of (we have some).

> I personally don't contribute very often different open source projects because I consider the overhead of setting up a development environment, debugging, and the possibility of submitting something that may be not accepted because done in the way the project administrators expected.

This is one of the reason we encourage people to discuss things before starting any work. Which is true for a lot of projects.

File: scipy/optimize/tests/test__shgo.py [hunk 1]
```diff
@@ -269,6 +269,12 @@ def g4(x):
     g = (g1, g2, g3, g4)
     cons = wrap_constraints(g)
 
+class StructTestExtraArgs(StructTestFunction):
+    def f(self, x, cx, cy):
+        return (x[0] - cx) ** 2 + (x[1] - cy) ** 2
+    g = None
+    cons = wrap_constraints(g)
+
```
Comment (mdhaber) [inline scipy/optimize/tests/test__shgo.py:277]: This looks unused, so I'll remove it.

**Comment by mdhaber** [comment]
I think I'll go ahead and merge if CI passes.

**Comment by mdhaber** [comment]
Thanks @o-alexandre-felipe @jjramsey! 
@Stefan-Endres sorry about the merge conflicts, but when you break up gh-16313, feel free to refactor this part as you see fit.

**Pull request merged**